control-plane/agents/core/src/volume/specs.rs

```diff
@@ -17,113 +17,105 @@ use common_lib::{
     mbus_api::ResourceKind,
     types::v0::{
         message_bus::{
-            ChildUri, CreateNexus, CreateReplica, CreateVolume, DestroyNexus, DestroyReplica,
-            DestroyVolume, Nexus, NexusId, NodeId, PoolState, Protocol, PublishVolume, ReplicaId,
-            ReplicaOwners, ShareNexus, ShareVolume, UnpublishVolume, UnshareNexus, UnshareVolume,
-            Volume, VolumeId, VolumeState,
+            AddNexusReplica, ChildUri, CreateNexus, CreateReplica, CreateVolume, DestroyNexus,
+            DestroyReplica, DestroyVolume, Nexus, NexusId, NodeId, Protocol, PublishVolume,
+            RemoveNexusReplica, Replica, ReplicaId, ReplicaOwners, SetVolumeReplica, ShareNexus,
+            ShareVolume, UnpublishVolume, UnshareNexus, UnshareVolume, Volume, VolumeId,
+            VolumeState,
         },
         store::{
-            nexus::NexusSpec,
+            nexus::{NexusSpec, ReplicaUri},
+            nexus_child::NexusChild,
             replica::ReplicaSpec,
             volume::{VolumeOperation, VolumeSpec},
             SpecState, SpecTransaction,
         },
     },
 };
+use parking_lot::Mutex;
 use snafu::OptionExt;
+use std::{convert::From, ops::Deref, sync::Arc};
 
-async fn get_node_pools(
+/// Select a nexus child to be removed from a nexus
+pub(crate) async fn get_nexus_child_remove_candidate(
+    spec: &VolumeSpec,
+    status: &Volume,
     registry: &Registry,
-    request: &CreateVolume,
-) -> Result<Vec<Vec<PoolWrapper>>, SvcError> {
-    let node_pools = registry.get_node_pools_wrapper().await?;
-
-    let size = request.size;
-    let replicas = request.replicas;
-    let allowed_nodes = request.allowed_nodes();
-
-    if !allowed_nodes.is_empty() && replicas > allowed_nodes.len() as u64 {
-        // oops, how would this even work mr requester?
-        return Err(SvcError::InvalidArguments {});
+) -> Result<ReplicaItem, SvcError> {
+    let candidates = scheduling::get_nexus_child_remove_candidate(
+        &GetChildForRemoval::new(spec, status),
+        registry,
+    )
+    .await;
+    tracing::trace!(
+        "Removal candidates for volume '{}': {:?}",
+        spec.uuid,
+        candidates
+    );
+
+    if candidates.len() <= 1 {
+        Err(SvcError::LastReplica {
+            id: "a".to_string(),
+        })
```

Comment:
Nit: If the number of candidates is less than 1 it isn't the last replica - there aren't any.

Reply:
Added new error: ReplicaRemovalNoCandidates